mltensor-explainer.md

```diff
@@ -0,0 +1,423 @@
+
+- Two fences are created:
+  1. a "start access" fence which is to be signaled by WebNN and waited on by WebGPU. A data copy may be required alongside the signaling of this fence
+  2. an "end access" fence which is to be signaled by WebGPU and waited on by WebNN. A data copy may be required alongside the signaling of this fence
+- The `GPUDevice` enqueues a command to its `GPUQueue` to wait for the "start access" fence to be signaled
+- WebNN will signal the "start access" fence after the completion of all currently-enqueued operations that use the `MLTensor` which is to be imported (this is very similar to how [`GPUBuffer.mapAsync()`](https://www.w3.org/TR/webgpu/#dom-gpubuffer-mapasync) works)
+- Until the "end access" fence is signaled:
+  - The `GPUDevice` has exclusive, read/write access to the imported buffer
+  - All WebNN work involving the imported `MLTensor` is blocked
+- When the `GPUBuffer` is destroyed, the "end access" fence is signaled and the `MLTensor` may be used again by WebNN
+
+### `compute()` vs. `dispatch()`
+
+`compute()` will be deprecated and removed in favor of `dispatch()`.
+
+It's possible `compute()` may have a performance advantage on some platforms for some use cases, such as in one-shot inference (where the benefits of buffer re-use are not relevant) with small inputs/outputs on CPU (where the overhead of task queueing may outweigh the benefits of parallelization). At this time, we do not expect this performance impact to be substantial enough to justify providing two mostly-identical methods for executing an `MLGraph`.
+
+### Open Questions
+
+- How will errors be surfaced? Do we need a concept similar to [WebGPU's error scopes](https://www.w3.org/TR/webgpu/#error-scopes), or is [returning errors via a promise for select operations sufficient](https://github.com/webmachinelearning/webnn/issues/697#issuecomment-2195656878)? See [#477](https://github.com/webmachinelearning/webnn/issues/477)
+- On non-UMA systems, does the user agent have enough information to appropriately allocate an `MLTensor` if an `MLDeviceType` is not used for creating an `MLContext`? See [#350](https://github.com/webmachinelearning/webnn/issues/350) and [#749](https://github.com/webmachinelearning/webnn/issues/749)
+- Should the `dispatch()` method be a part of the `MLGraph` interface rather than `MLContext`? Should `readBuffer()` and `writeBuffer()` exist on an `MLTensor`? See [#697](https://github.com/webmachinelearning/webnn/issues/697).
+- If an `MLContext` is not created from a `GPUDevice`, does there need to be some mechanism - above and beyond the `MLTensorUsage.WEBGPU_INTEROP` flag - for identifying the specific `GPUDevice` with which interop is desired?
```

Comment:
Since the WebGPU Import method is a member of GPUDevice, isn’t the target destination for the MLTensor simply the GPUDevice on which Import was called?

Reply:
Thanks for calling this out because the question as written is not the most clear.

The `MLTensor` is allocated before it is imported to WebGPU. On which GPU should this tensor be allocated?

Reply:
I think the answer is that it should always be allocated on the device chosen for execution with MLContext. 

When I've spoken with @RafaelCintron, the plan for cross device interop is to have WebNN copy the WebNN resource onto an intermediate resource located on the common device and then use that common resource with WebGPU interop as usual (and copy it back upon return to WebNN). This object _could_ be allocated at MLTensor creation time, however I expect we'd do it lazily upon MLTensor import.

The tradeoff of always allocating on the WebNN device is that if a user decides creates the MLTensor and intends to access it with WebGPU first, there would be the overhead of that additional WebNN to common resource sync, but really we could optimize out this operation (since there's no real data to by synced).

Reply:
Removed this open question because yeah this is just another case of cross-device interop. It would be unfortunate to allocate the tensor on the "wrong" GPU, but then again the easy way for the developer to protect against this is to create the `MLContext` with the `GPUDevice` they want to use for interop